astropy/units/physical.py

```diff
@@ -1,133 +1,536 @@
+        new_unit = getattr(self._unit, operation)(other_unit)
+        return new_unit.physical_type
+
+    def __mul__(self, other):
+        return self._dimensional_analysis(other, "__mul__")
+
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    def __truediv__(self, other):
+        return self._dimensional_analysis(other, "__truediv__")
+
+    def __rtruediv__(self, other):
+        other = self._dimensionally_compatible_unit(other)
+        if other is NotImplemented:
+            return NotImplemented
+        return other.physical_type._dimensional_analysis(self, "__truediv__")
+
+    def __pow__(self, power):
+        return (self._unit ** power).physical_type
+
+    def __hash__(self):
+        return hash(self._physical_type_id)
+
+    def __len__(self):
+        return len(self._physical_type)
+
+    # We need to prevent operations like where a Unit instance left
+    # multiplies a PhysicalType instance from returning a `Quantity`
+    # instance with a PhysicalType as the value.  We can do this by
+    # preventing np.array from casting a PhysicalType instance as
+    # an object array.
+    __array__ = None
+
+
+def _get_names_in_use_except_from(unit):
+    """
+    Get a `set` containing all of the physical type names in use, except
+    for the names that correspond to ``unit``.
+    """
+    names_in_use = set(_unit_physical_mapping.keys())
+    id = unit._get_physical_type_id()
+    known_physical_type = id in _physical_unit_mapping.keys()
+    names_for_unit = set(_physical_unit_mapping[id]) if known_physical_type else set()
+    return names_in_use - names_for_unit
 
 
 def def_physical_type(unit, name):
     """
-    Adds a new physical unit mapping.
+    Add a mapping between a unit and the corresponding physical type(s).
+
+    If a physical type already exists for a unit, add new physical type
+    names so long as those names are not already in use for other
+    physical types.
 
     Parameters
     ----------
-    unit : `~astropy.units.UnitBase` instance
-        The unit to map from.
+    unit : `~astropy.units.Unit`
+        The unit to be represented by the physical type.
+
+    name : `str` or `set` of `str`
+        A `str` representing the name of the physical type of the unit,
+        or a `set` containing strings that represent one or more names
+        of physical types.
 
-    name : str
-        The physical name of the unit.
+    Raises
+    ------
+    ValueError
+        If a physical type name is already in use for another unit, or
+        if attempting to name a unit as ``"unknown"``.
     """
-    r = unit._get_physical_type_id()
-    if r in _physical_unit_mapping:
+    physical_type_names = _standardize_physical_type_names(name)
+
+    if "unknown" in physical_type_names:
+        raise ValueError("cannot uniquely define an unknown physical type")
+
+    names_for_other_units = _get_names_in_use_except_from(unit)
```

Comment:
Actually, I think this could just become a one-liner instead of an additional function.